Merge Plotting guide into document_sprint branch (#366)
columnflow/columnflow: docs/user_guide/plotting.md

# Plotting

In columnflow, there are multiple tasks to create plots. This section showcases, how to create and
customize a plot based on the {py:class}`~columnflow.tasks.plotting.PlotVariables1D` task. A
detailed overview of all plotting tasks is given in the [Plotting tasks](../task_overview/plotting.md) section.

## Creating your first plot

Assuming you used the analysis template to setup your analysis, you can create a first plot by running
```shell
law run cf.PlotVariables1D --version v1 \
    --calibrators example --selector example --producer example \
    --processes data,tt,st --variables n_jet --categories incl,2j
```
This will run the full analysis chain for the given processes (data, tt, st) and should create
plots looking like this:

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__12dfac316a__plot__proc_3_7727a49dc2__cat_incl__var_n_jet.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__12dfac316a__plot__proc_3_7727a49dc2__cat_2j__var_n_jet.pdf
:width: 100%
:::
::::

> 

Author: For me, the plots are directly visible in the documentation. No idea why it's different for you.
![Screenshot 2023-12-21 at 10 14 25](https://github.com/columnflow/columnflow/assets/49306645/c14edfe5-248a-4528-9247-88d3fcce6ca1)

Regarding the `1e6`, we should fix that in a separate PR. I opened an Issue for that (https://github.com/columnflow/columnflow/issues/373)

the runtime of the task, e.g. via ```--view-cmd evince-previewer```.
:::

The PlotVariables1D task is located at the bottom of our
[task graph](https://github.com/columnflow/columnflow/wiki#default-task-graph), which means that
all tasks leading to PlotVariables1D will be run for all datasets corresponding to the
```--processes``` we requested using the Calibrators, Selector, and Producers
(often referred to as CSPs) as requested. In the following examples, we will skip the
```--calibrators```, ```selector``` and ```producers``` parameters, which means that the defaults
defined in the config will be used automatically. Examples on how to
implement your own CSPs can be found in the [calibrators](building_blocks/calibrators),
[selectors](building_blocks/selectors), and [producers](building_blocks/producers) sections of the
user guide. The ```--variables``` parameter defines, for which variables we want to create histograms
and plots. Variables are order objects that need to be defined in the config as shown in the
[config objects](building_blocks/config_objects) section and the column corresponding to the expression

> statement needs to be stored either after the {py:class}`~columnflow.tasks.reduction.ReduceEvents`

Author: added :+1:

For each of the category given with the ```--categories``` parameter, one plot will be produced.
A detailed guide on how to implement categories in Columnflow is given in the
[categories](building_blocks/categories) section.

To define, which processes and datasets to consider when plotting, you can use the ```--processes```

> and ```--datasets``` parameter. When only processes are given, all datasets corresponding to the

Author: I'm not sure if a agree, because we use the words datasets and processes quite often.
I'd only highlight processes/datasets, if I'm actually talking about a `Process` object or about the `--processes` parameter

requested processes will be considered. When only datasets are given, all processes in the config
will be considered.
The ```processes``` parameter can be used to change the order of processes in the stack and the legend
(try for example ```--processes st,tt``` instead) and to further distinguish between sub-processes
(e.g. via ```--processses, tt_sl,tt_dl,tt_fh```).

:::{dropdown} Note! Do not add the same dataset via multiple processes!
At the time of writing this documentation, there is still an issue present that histograms corresponding
to a dataset can accidentally be used multiple times. For example, when adding ```--processes tt,tt_sl```,
the events corresponding to the dataset ```tt_sl_powheg``` will be displayed twice in the resulting
plot.
:::

> ## Customization of plots

Reviewer: The customiztion of plots is also applicable for 2D and cutflow plots, right? If this is correct, we might want to consider moving this section to the end of the page

Author: Yes, the parameters are mostly the same between tasks, but my idea for the guide was to first present all the customization options based on the PlotVariables1D task and only afterwards present the other types of tasks. But I can add one or two sentences to mention that the parameters are mostly the same between tasks.

Author: true, added now

There are many different parameters implemented that allow customizing the style of a plot. A short
description to all plotting parameters is given in the [Plotting tasks](../task_overview/plotting.md).
In the following, a few exemplary task calls are given to present the usage of our plotting parameters.

> Per default, the PlotVariables1D task creates one plot per variable with all Monte Carlo backgrounds being included

Author: This is handled by the `is_mc`/`is_data` attribute of the dataset, but I don't think this documented anywhere yet.

> Per default, the PlotVariables1D task creates one plot per variable with all Monte Carlo backgrounds being included

Author: Good point, changed :+1:

in a stack and data being shown as separate points. The bottom subplot shows the ratio between signal
and all processes included in the stack and can be disabled via the ```--skip_ratio``` parameter.
To change the text next to the label, you can add the ```--cms-label``` parameter.

> To compare shapes of multiple processes, you might want to plot each process separately as one line.

Reviewer: It might be good to use separate paragraphs for different style options

Author: I feel like having a paragraph for each parameter might be overkill since some parameters are just mentioned in a single sentence. Maybe one new paragraph before discussing the `process-settings` parameter?

To achieve this, you can use the ```unstack``` option of the ```--process-settings``` parameter. This
parameter can also be used to change other attributes of your process instances, such as color, label,
and the scale. To better compare shapes of processes, we can normalize each line with the
```--shape-norm``` parameter. Combining all the previously discussed parameters might lead to a task
call such as

```shell
law run cf.PlotVariables1D --version v1 --processes tt,st --variables n_jet,jet1_pt \
    --skip-ratio --shape-norm --cms-label simpw \
    --process-settings "tt,unstack,color=#e41a1c:st,unstack,label=Single Top"

> ```

Reviewer: The `--cms-label simpw` might be a little misleading since we haven't discussed the keywords for the cms-label here. We might want to include e.g. a table of keywords and how they are translated for the final plot

Author: good point, added with the next commit

to produce the following plot:

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__0191de868f__plot__proc_2_a2211e799f__cat_incl__var_jet1_pt__c1.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__0191de868f__plot__proc_2_a2211e799f__cat_incl__var_n_jet__c1.pdf
:width: 100%
:::
::::

Parameters that only contain a single value can also be passed via the ```--general-settings```.
We can also change the y-scale of the plot to a log scale by adding ```--yscale log``` and change some

> properties of specific variables via the ```variable-settings``` parameter. An exemplary task call

Reviewer: I think explaining a little more what we want to do in this example might be instructive. Maybe sth like

```suggestion
properties of specific variables via the ```variable-settings``` parameter. For example, we might want to create the plots of our two obserables in one call, but would like to try out a rebinned version of `jet1_pt` that only consideres every tenth bin. A corresponding task call
```

might be

```shell
law run cf.PlotVariables1D --version v1 --processes tt,st --variables n_jet,jet1_pt \

>     --general-settings "skip_ratio,shape_norm,yscale=log,cms-label=simpw" \

Reviewer: Stupid question, but why do you have to use the `--general-settings` option in the first place? Can't you just use

```
--skip-ratio --shape-norm --yscale log --cms-label simpw
```

instead?

Author: Yes, they are producing the same result, but I wanted to showcase how to use the general_settings.

Reviewer: OK, I see your point and generally agree. But as a reader, I would be wondering _why_ I need the general_settings in the first place if I can do the same with the command line options. I think adding some explanation(s) about when to use which of these ways of parsing options would be beneficial

Author: added :+1:

    --variable-settings "n_jet,y_title=Events,x_title=N jets:jet1_pt,rebin=10,x_title=Leading jet \$p_{T}\$"
```

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__c80529af83__plot__proc_2_a2211e799f__cat_incl__var_jet1_pt__c2.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__c80529af83__plot__proc_2_a2211e799f__cat_incl__var_n_jet__c2.pdf
:width: 100%
:::
::::

:::{dropdown} Limitations of the ```variable_settings```

> While in theory, we can change anything inside the variable and process instances via the

Reviewer: ```suggestion
While in theory we can change anything inside the variable and process instances via the
```

```variable_settings```, there are certain attributes that are already used during the creation

> of the histograms (e.g. the ```expression``` and the ```binning```) and since our ```variable_settings```

Reviewer: ```suggestion
of the histograms (e.g. the ```expression``` and the ```binning```). Since our ```variable_settings```
```

```

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__be60d3bca7__plot__proc_2_a2211e799f__cat_incl__var_jet1_pt__c3.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotVariables1D_tpl_config_analy__1__be60d3bca7__plot__proc_2_a2211e799f__cat_incl__var_n_jet__c3.pdf
:width: 100%
:::
::::

## Creating 2D plots

Columnflow also provides the {py:class}`~columnflow.tasks.plotting.PlotVariables2D` to create
two-dimensional plots. Two-dimensional histograms are created by passing two variables to the
```--variables``` parameter, separated by a ```-```. Here is an exemplary task call and their
outputs.
```shell
law run cf.PlotVariables2D --version v1 \
    --processes tt,st --variables n_jet-jet1_pt,jet1_pt-n_jet
```

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotVariables2D_tpl_config_analy__1__b27b994979__plot__proc_2_a2211e799f__cat_incl__var_jet1_pt-n_jet.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotVariables2D_tpl_config_analy__1__b27b994979__plot__proc_2_a2211e799f__cat_incl__var_n_jet-jet1_pt.pdf
:width: 100%
:::
::::

For more information on the task parameters of the ```PlotVariables2D``` task, take a look into the

> [plotting task overview](../task_overview/plotting.md).

Reviewer: Maybe mention here that many (if not all!) options used in PlotVariables1D can also be used for the 2D plots

## Creating cutflow plots

The previously discussed plotting functions only create plots after applying the full event selection.
To allow inspecting and optimizing an event and object selection, Columnflow also includes plotting

> tasks that can be run before applying any event selections.

Reviewer: ```suggestion
tasks that can be run before applying any or a given set of event selections.
```

Author: This suggestion looks a bit confusing to me. Maybe something like this?
```suggestion
tasks that can produce plots after each individual selection step.
```

Reviewer: LGTM

> To create a simple cutflow plots, displaying event yields after each individual selection step,

Reviewer: ```suggestion
To create a simple cutflow plot, displaying event yields after each individual selection step,
```

you can use the {py:class}`~columnflow.tasks.cutflow.PlotCutflow` task, e.g. via calling
```shell
law run cf.PlotCutflow --version v1 \
    --calibrators example --selector example --categories incl,2j \
    --shape-norm --process-settings tt,unstack:st,unstack \
    --processes tt,st --selector-steps jet,muon

> ```

Reviewer: This call also assumes that you actually defined selection steps called `jet` and `muon` in a Selector `example`. Might make sense to explicitely write that here again

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotCutflow_tpl_config_analy__1__12a17bf79c__cutflow__cat_incl.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotCutflow_tpl_config_analy__1__12a17bf79c__cutflow__cat_2j.pdf
:width: 100%
:::
::::

This will produce a plot with three bins, containing the event yield before applying any selection

> and after each selector step, where we always apply the locical and of all previous selector steps.

Reviewer: ```suggestion
and after each selector step, where we always apply the logical ```and``` of all previous selector steps.
```

To create plots of variables as part of the cutflow, we also provide the
{py:class}`~columnflow.tasks.cutflow.PlotCutflowVariables1D`, which mostly behaves the same as the
PlotVariables1D task.

```shell
law run cf.PlotCutflowVariables1D --version v1 \
    --calibrators example --selector example \
    --processes tt,st --variables cf_jet1_pt --categories incl \
    --selector-steps jet,muon --per-plot processes
```

::::{grid} 1 1 3 3
:::{figure} ../plots/cf.PlotCutflowVariables1D_tpl_config_analy__1__d8a37d3da9__plot__step0_Initial__proc_2_a2211e799f__cat_incl__var_cf_jet1_pt.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotCutflowVariables1D_tpl_config_analy__1__d8a37d3da9__plot__step1_jet__proc_2_a2211e799f__cat_incl__var_cf_jet1_pt.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotCutflowVariables1D_tpl_config_analy__1__d8a37d3da9__plot__step2_muon__proc_2_a2211e799f__cat_incl__var_cf_jet1_pt.pdf
:width: 100%
:::
::::

The ```per-plot``` parameter defines whether to produce one plot per selector step

> (```per-plot processes```) or one plot per process (```per-plot steps```).

Reviewer: This seems very counter intuitive. Are you sure this isn't switched around?

Author: I agree that it's confusing, but that's how it is implemented. `per-plot processes` means that you get all processes per plot.

Reviewer: OK then :D

For the ```per-plot steps``` option, try the following task call:

```shell
law run cf.PlotCutflowVariables1D --version v1 \
    --calibrators example --selector example \
    --processes tt,st --variables cf_jet1_pt --categories incl \
    --selector-steps jet,muon --per-plot steps
```

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotCutflowVariables1D_tpl_config_analy__1__c3947accbb__plot__proc_st__cat_incl__var_cf_jet1_pt.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotCutflowVariables1D_tpl_config_analy__1__c3947accbb__plot__proc_tt__cat_incl__var_cf_jet1_pt.pdf
:width: 100%
:::
::::

## Creating plots for different shifts

Like most tasks, our plotting tasks also contain the ```--shift``` parameter that allows requesting

> the outputs for a certain type of systematic variation. Per default, the ```shellift```parameter is set

Reviewer: ```suggestion
the outputs for a certain type of systematic variation. Per default, the ```shift``` parameter is set
```

?

to "nominal", but you could also produce your plot with a certain systematic uncertainty varied
up or down, e.g. via running
```shell
law run cf.PlotVariables1D --version v1 \
    --processes tt,st --variables n_jet --shift mu_up
```

If you already ran the same task call with ```--shift nominal``` before, this will only require to
produce new histograms and plots, as a shift such as the ```mu_up``` is typically implemented as an
event weight and therefore does not require to reproduce any columns. Other shifts such as ```jec_up```
also impact our event selection and therefore also need to re-run anything starting from
{py:class}`~columnflow.tasks.selection.SelectEvents`. A detailed overview on how to implement different
types of systematic uncertainties is given in the [systematics](systematics)
section (TODO: not existing).

For directly comparing differences introduced by one shift source, we provide the
{py:class}`~columnflow.tasks.plotting.PlotShiftedVariables1D` task. Instead of the ```--shift```

> parameter, this task implements the ```--shift-sources``` task and creates one plot per shift source

Reviewer: ```suggestion
parameter, this task implements the ```--shift-sources``` option and creates one plot per shift source
```

right?

displaying the nominal distribution (black) compared to the shift source varied up (red) and down (blue).
The task can be called e.g. via
```shell
law run cf.PlotShiftedVariables1D --version v1 \
    --processes tt,st --variables jet1_pt,n_jet --shift-sources mu
```
and produces the following plot:

::::{grid} 1 1 2 2
:::{figure} ../plots/cf.PlotShiftedVariables1D_tpl_config_analy__1__42b45aba89__plot__proc_2_a2211e799f__unc_mu__cat_incl__var_jet1_pt.pdf
:width: 100%
:::

:::{figure} ../plots/cf.PlotShiftedVariables1D_tpl_config_analy__1__42b45aba89__plot__proc_2_a2211e799f__unc_mu__cat_incl__var_n_jet.pdf
:width: 100%
:::
::::

This produces per default only one plot containing the sum of all processes. To produce this plot
per process, you can use the {py:class}`~columnflow.tasks.plotting.PlotShiftedVariablesPerProcess1D`
task

## Directly displaying plots in the terminal

All plotting tasks also include a ```--view-cmd``` parameter that allows directly printing the plot
during the runtime of the task:
```shell
law run cf.PlotVariables1D --version v1 \
    --processes tt,st --variables n_jet --view-cmd evince-previewer
```

> ## Using you own plotting function

Reviewer: ```suggestion
## Using your own plotting function
```

> While all plotting tasks provide default plotting functions, which implement many parameters to

Reviewer: ```suggestion
While all plotting tasks provide default plotting functions which implement many parameters to
```
